go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go

```diff
@@ -356,14 +365,44 @@ func (ts *txThrottlerStateImpl) throttle() bool {
 	// Serialize calls to ts.throttle.Throttle()
 	ts.throttleMu.Lock()
 	defer ts.throttleMu.Unlock()
-	return ts.throttler.Throttle(0 /* threadId */) > 0
+
+	maxLag := atomic.LoadInt64(&ts.shardMaxLag)
+
+	return maxLag > ts.config.TxThrottlerConfig.TargetReplicationLagSec &&
+		ts.throttler.Throttle(0 /* threadId */) > 0
+}
+
+func (ts *txThrottlerStateImpl) updateMaxShardLag() {
+	defer ts.endWaitGroup.Done()
+	// We use half of the target lag to ensure we have enough resolution to see changes in lag below that value
+	ticker := time.NewTicker(time.Duration(ts.config.TxThrottlerConfig.TargetReplicationLagSec/2) * time.Second)
+	defer ticker.Stop()
+outerloop:
+	for {
+		select {
+		case <-ticker.C:
+			var maxLag uint32
+
+			for tabletType := range ts.tabletTypes {
+				maxLagPerTabletType := ts.throttler.LastMaxLagNotIgnoredForTabletType(tabletType)
+				if maxLagPerTabletType > maxLag {
+					maxLag = maxLagPerTabletType
+				}
+			}
+			atomic.StoreInt64(&ts.shardMaxLag, int64(maxLag))
+		case <-ts.endChannel:
+			break outerloop
+		}
+	}
 }
 
 func (ts *txThrottlerStateImpl) deallocateResources() {
 	// Close healthcheck and topo watchers
 	ts.closeHealthCheckStream()
 	ts.healthCheck = nil
 
+	ts.endChannel <- true
+	ts.endWaitGroup.Wait()
```

Comment:
Do we really need a wait group for this? It seems to me that once you send a message on `endChannel`, `updateMaxShardLag` is guaranteed to complete, and stop the ticker, which is really the only resource we are worried about releasing.

Reply:
Ugh. I guess even checking for nil before accessing `ts.throttler` leaves a tiny slice of time where you could still get a panic. It's fine for now.